lib/components.js

```diff
@@ -6,6 +6,231 @@
 const newLinesRegExp = new RegExp('^\s+|\s+$', 'g');
 
 const components = function () {
+    // This is the base class that holds all functionality shared between custom components
+    // and native elements
+    class Validator {
+        /**
+         * Check if element if child of a form
+         * @param {HTMLElement} element
+         * @returns {boolean}
+         */
+        static isFormElement(element) {
+            element = element.parentElement;
+            while (element && element.parentElement) {
+                if (element.tagName === 'GAMEFACE-FORM-CONTROL' || element.tagName === 'gameface-form-control') return true;
+                element = element.parentElement;
+            }
+
+            return false;
+        }
+
+        /**
+         * Most of the custom elements will not need this check however,
+         * we call all validation methods in order to determine if an element is valid.
+         * Each element that needs this check implements it itself.
+         * @returns {boolean}
+         */
+        static tooLong() {
+            return false;
+        }
+
+        /**
+        * Most of the custom elements will not need this check however,
+        * we call all validation methods in order to determine if an element is valid.
+        * Each element that needs this check implements it itself.
+        * @returns {boolean}
+        */
+        static tooShort() {
+            return false;
+        }

```

Comment:
We have it because in the hasErrors function in the form control we call all check functions:
```
const errorNames = {
            notAForm: !element.isFormElement(),
            tooLong: element.tooLong(),
            tooShort: element.tooShort(),
            rangeOverflow: element.rangeOverflow(),
            rangeUnderflow: element.rangeUnderflow(),
            valueMissing: element.valueMissing(),
            nameMissing: element.nameMissing(),
            customError: element.customError()
        };
```

And if some of these doesn't have implementation this check will throw error.
They handles the cases where an tooShort, tooLong etc are not applicable to certain element.
Tried to apply the same implementation that Chromium has:
1. [call all validation methods](https://github.com/chromium/chromium/blob/9e582c8c1e0d72eb68bdebe2d33de91c5a4282a4/third_party/blink/renderer/core/html/forms/listed_element.cc#L385)
2. [have implementations for all validation methods](https://github.com/chromium/chromium/blob/9e582c8c1e0d72eb68bdebe2d33de91c5a4282a4/third_party/blink/renderer/core/html/forms/listed_element.cc#L353)
3. [implement custom validations in the specific component](https://github.com/chromium/chromium/blob/2c1b436d68be93925325da5748a663d3beaf6ccc/third_party/blink/renderer/core/html/forms/input_type.cc#L232)